File: bindings/rust/perf/multi/src/perf_multi_dealer_dealer_server.rs
//! DEALER-DEALER multi server: one-way active receiver.

#[path = "perf_common.rs"]
mod common;

use std::io::{self, BufRead};
use std::time::{Duration, Instant};
use zlink::{POLLIN, PollEvent, Poller, Received, RecvFlags, RecvResult};

fn main() {
    let args = common::MultiArgs::parse();
    let settings = common::MultiSettings::from_env();

    let ctx = common::perf_server_context();
    common::apply_multi_auto_hwm_msg_unit(&ctx, args.msg_size);
    let server = ctx.dealer_socket().expect("dealer");
    // C parity: numeric HWM remains behind the manual-override gate.
    common::apply_multi_hwm(&server, &settings);
    server
        .common_options()
        .set_receive_timeout(Duration::from_millis(settings.receive_timeout_ms))
        .expect("rcvtimeo");
    if matches!(args.transport.as_str(), "tls" | "wss") {
        let tls = common::resolve_perf_tls_paths().expect("TLS certs not found");
        common::setup_raw_tls_server(&server, &tls).expect("server tls");
    }

    let Some(bind_endpoint) =
        common::resolve_server_bind_endpoint("MULTI_DEALER_DEALER", &args.transport)
    else {
        return;
    };
    if let Err(err) = server.bind(&bind_endpoint) {
        if common::handle_transport_setup_error("MULTI_DEALER_DEALER", &args.transport, "bind", err)
        {
            return;
        }
        panic!("bind: {err}");
    }
    let endpoint = server.last_endpoint().expect("endpoint");
    common::print_ready(&endpoint);

    let stdin = io::stdin();
    let mut start_seen = false;
    for line in stdin.lock().lines() {
        let line = line.unwrap_or_default();
        if line.trim() == format!("START,{}", args.msg_size) {
            start_seen = true;
            break;
        }
        if matches!(line.trim(), "STOP" | "QUIT") {
            return;
        }
    }
    if !start_seen {
        return;
    }

    // C perf_multi_dealer_dealer_server.cpp run_receive_window(): poller POLLIN,
    // perf_socket_poll(...,-1) signal-driven, drain non-blocking, count until
    // the measure-seconds deadline. The wire stop token only wakes the poll; it
    // does NOT anchor aggregation (the server deadline does).
    let deadline = Instant::now() + Duration::from_secs(settings.duration_seconds.max(1));
    let mut latency_stats = common::LatencyStats::new();
    let mut active_count: u64 = 0;
    let mut received = Received::empty();

    let poller = Poller::new().expect("poller");
    poller.add_socket(&server, POLLIN, 0).expect("poller add");
    let mut events = vec![PollEvent::default(); 1];

    let mut record = |received: &Received, latency_stats: &mut common::LatencyStats| {
        let data = common::message_payload(received.parts());
        // Stop token (is_valid_active_message is false for it) only ends the
        // window via the deadline; it is never counted.
        if Instant::now() < deadline && common::is_valid_active_message(data, args.msg_size) {
            let sent_ts_ns = common::decode_sent_ts_ns(data);
            latency_stats
                .record_ns(common::now_ns().saturating_sub(sent_ts_ns.max(0) as u64) as f64);
            active_count += 1;
        }
    };

    'outer: loop {
        let remaining_ms = deadline
            .saturating_duration_since(Instant::now())
            .as_millis()
            .max(1) as i64;
        match poller.wait(&mut events, remaining_ms) {
            Ok(n) if n > 0 && events[0].is_readable() => {}
            Ok(_) if Instant::now() < deadline => continue,
            Ok(_) => break 'outer,
            Err(err) => panic!("poller wait failed: {err}"),
        }
        // Drain everything currently queued without blocking.
        loop {
            match server.recv(&mut received, RecvFlags::DONT_WAIT) {
                Ok(true) => {
                    record(&received, &mut latency_stats);
                }
                Ok(false) => break,
                Err(err) if err.code() == RecvResult::NoData => break,
                Err(err) => panic!("recv failed: {err}"),
            }
        }
        if Instant::now() >= deadline {
            break 'outer;
        }
    }

    // C drain_phase_until_idle(2.0s max, 50ms idle): UNCOUNTED tail drain so
    // stale messages do not spill into the next size case.
    let drain_deadline = Instant::now() + Duration::from_secs_f64(2.0);
    let mut idle_deadline = Instant::now() + Duration::from_millis(50);
    while Instant::now() < drain_deadline {
        match server.recv(&mut received, RecvFlags::DONT_WAIT) {
            Ok(true) => {
                idle_deadline = Instant::now() + Duration::from_millis(50);
            }
            Ok(false) | Err(_) => {
                if Instant::now() >= idle_deadline {
                    break;
                }
                common::poll_idle_until(idle_deadline, Duration::from_millis(1));
            }
        }
    }

    let stats = latency_stats.finish();
    let final_stats = common::StatsResult {
        count: active_count,
        ..stats
    };
    common::print_result(
        "MULTI_DEALER_DEALER",
        &args.transport,
        args.msg_size,
        settings.duration_seconds,
        &final_stats,
    );
}

File: bindings/python/perf/multi/perf_multi_dealer_dealer_server.py
import sys
import threading
import time

import zlink

from perf_multi_common import (
    active_message_latency_ns,
    apply_multi_socket_options,
    benchmark_endpoint,
    benchmark_run_id,
    configure_multi_tls_server,
    LatencySampler,
    print_result_lines,
    parse_server_args,
    perf_server_context,
    resolve_multi_monitor_hwm_bytes,
    result_metrics,
    received_has_stop_token,
    received_metric_payload,
    safe_poll,
)


def dealer_dealer_active_poll_timeout_ms(
    stop_token_seen, active_deadline, *, now=None
):
    if not stop_token_seen:
        return -1
    current = time.perf_counter() if now is None else now
    return max(1, int((active_deadline - current) * 1000))


def main(argv=None):
    args = parse_server_args(argv or sys.argv[1:])
    run_id = benchmark_run_id()
    endpoint = benchmark_endpoint(args.transport, "multi-dealer-dealer")
    active_duration_s = max(1.0, float(args.duration))
    start_event = threading.Event()
    stop_event = threading.Event()

    def read_commands():
        for line in sys.stdin:
            text = line.strip().upper()
            if text == f"START,{args.msg_size}":
                start_event.set()
            elif text in {"STOP", "QUIT"}:
                stop_event.set()
                start_event.set()
                return

    threading.Thread(target=read_commands, daemon=True).start()

    with perf_server_context() as ctx:
        with zlink.create_dealer_socket(ctx) as dealer:
            configure_multi_tls_server(dealer, args.transport)
            apply_multi_socket_options(dealer)
            with dealer.monitor_open(
                zlink.MonitorEventMask.CONNECTION_READY,
                resolve_multi_monitor_hwm_bytes(),
            ) as monitor:
                dealer.bind(endpoint)
                print(f"READY,{endpoint}", flush=True)
                with zlink.create_poller() as poller:
                    poller.add_socket(dealer, zlink.PollEventFlag.POLLIN, 0)
                    poll_events = zlink.create_poll_events(1)
                    start_event.wait()
                    if stop_event.is_set():
                        return
                    active_deadline = time.perf_counter() + active_duration_s
                    latency_sampler = LatencySampler()
                    count = 0
                    stop_token_seen = False
                    recv_storage = zlink.create_received()
                    recv_into = dealer.recv_into
                    recv_flags = zlink.RecvFlags.DONT_WAIT

                    def drain_ready():
                        # C receive_one_message + drain_non_blocking_messages:
                        # every matched header counts; latency excludes
                        # clock-skew (latency_ns_from_message returns None).
                        nonlocal count, stop_token_seen
                        while True:
                            if not recv_into(recv_storage, flags=recv_flags):
                                return
                            with recv_storage:
                                if not recv_storage.parts:
                                    continue
                                if received_has_stop_token(recv_storage):
                                    stop_token_seen = True
                                    continue
                                data = received_metric_payload(
                                    recv_storage, expected_size=args.msg_size
                                )
                                if not data:
                                    continue
                                active, latency = active_message_latency_ns(
                                    data,
                                    expected_msg_size=args.msg_size,
                                    run_id=run_id,
                                )
                                if not active:
                                    continue
                                if time.perf_counter() >= active_deadline:
                                    continue
                                count += 1
                                if latency is not None:
                                    latency_sampler.add(latency)

                    while not stop_event.is_set():
                        now = time.perf_counter()
                        if now >= active_deadline:
                            break
                        # Match the C receiver: readiness is signal-driven
                        # until a valid wire stop token is consumed. Only
                        # then may the remaining active interval bound an
                        # otherwise empty wait.
                        poll_timeout_ms = dealer_dealer_active_poll_timeout_ms(
                            stop_token_seen,
                            active_deadline,
                            now=now,
                        )
                        ready_count = safe_poll(
                            poller, poll_events, poll_timeout_ms
                        )
                        if ready_count:
                            for offset in range(ready_count):
                                if poll_events.revents(offset) & int(
                                    zlink.PollEventFlag.POLLIN
                                ):
                                    drain_ready()
                        if time.perf_counter() >= active_deadline:
                            break

                    # C drain_phase_until_idle: bounded uncounted tail drain
                    # so late in-flight messages do not skew the next case.
                    idle_deadline = time.perf_counter() + 0.05
                    tail_deadline = time.perf_counter() + active_duration_s
                    while (
                        not stop_event.is_set()
                        and time.perf_counter() < tail_deadline
                    ):
                        if recv_into(recv_storage, flags=recv_flags):
                            with recv_storage:
                                pass
                            idle_deadline = time.perf_counter() + 0.05
                            continue
                        if time.perf_counter() >= idle_deadline:
                            break
                        safe_poll(poller, poll_events, 50)

                    if count <= 0:
                        raise RuntimeError(
                            "multi dealer-dealer server did not receive any active message"
                        )
                    if latency_sampler.count == 0:
                        raise RuntimeError(
                            "multi dealer-dealer server received active messages without latency samples"
                        )
                    metrics = result_metrics(
                        count=count,
                        msg_size=args.msg_size,
                        elapsed_s=args.duration,
                        latency_sampler=latency_sampler,
                    )
                    print_result_lines(
                        "MULTI_DEALER_DEALER",
                        args.transport,
                        args.msg_size,
                        metrics,
                    )


if __name__ == "__main__":
    main()
